### ISSUE TITLE
Job cancellation/retry is broken

### ISSUE THREAD

**Issue by Kamforka** [samuelcolvin/arq#190]
I've tried to reproduce the automatic job retry mechanism of `arq` based on [Retrying jobs and cancellation](https://arq-docs.helpmanual.io/#retrying-jobs-and-cancellation).

However whenever I press Ctrl+C  - once my task already started but before it's finished - the worker produces a nasty exception traceback like this:

```python3
13:14:31: Starting worker for 1 functions: long_running_task
13:14:31: redis_version=5.0.9 mem_usage=858.92K clients_connected=2 db_keys=15
13:14:33:   0.24s → caefef46a4fa43618ece3fd77f92cd2a:long_running_task()
Running...
^C13:14:34: shutdown on SIGINT ◆ 0 jobs complete ◆ 0 failed ◆ 0 retries ◆ 1 ongoing to cancel
Traceback (most recent call last):
  File "/home/s6dnau/git/hx-api/venv/bin/arq", line 10, in <module>
    sys.exit(cli())
  File "/home/s6dnau/git/hx-api/venv/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/s6dnau/git/hx-api/venv/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/s6dnau/git/hx-api/venv/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/s6dnau/git/hx-api/venv/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/s6dnau/git/hx-api/venv/lib/python3.8/site-packages/arq/cli.py", line 48, in cli
    run_worker(worker_settings_, **kwargs)
  File "/home/s6dnau/git/hx-api/venv/lib/python3.8/site-packages/arq/worker.py", line 630, in run_worker
    worker.run()
  File "/home/s6dnau/git/hx-api/venv/lib/python3.8/site-packages/arq/worker.py", line 235, in run
    self.loop.run_until_complete(self.close())
  File "/home/s6dnau/.pyenv/versions/3.8.0/lib/python3.8/asyncio/base_events.py", line 608, in run_until_complete
    return future.result()
asyncio.exceptions.CancelledError
```

Then I immediately restart my worker and the cancelled job is only retried after ~310 seconds:

```python3
13:14:37: Starting worker for 1 functions: long_running_task
13:14:37: redis_version=5.0.9 mem_usage=859.58K clients_connected=2 db_keys=20
13:19:43: 310.57s → caefef46a4fa43618ece3fd77f92cd2a:long_running_task() try=2 delayed=310.57s
Running...
Done!
13:19:48:   5.00s ← caefef46a4fa43618ece3fd77f92cd2a:long_running_task ● 
```

Why is that and how can I control the time delay before a job is retried automatically?

The only thing that is close to 310 seconds is the `job_timeout` parameter of the `Worker`, but in this context it doesn't make sense to me.

Here's a minimalistic snippet to reproduce my issue:
```python3
import asyncio

from arq import create_pool
from arq.connections import RedisSettings


async def long_running_task(ctx):
    print("Running...")
    await asyncio.sleep(5)
    print("Done!")


async def main():
    redis = await create_pool(RedisSettings())
    await redis.enqueue_job("long_running_task")


class WorkerSettings:
    functions = [long_running_task]


if __name__ == "__main__":
    loop = asyncio.get_event_loop()
    loop.run_until_complete(main())
```

System info:
```
os==ubuntu server 16.04
python==3.8.1
arq==0.19
```

**Comment by Kamforka** [comment]
So yeah I set the `job_timeout` to `10.0` and now the job retry was delayed by ~20 seconds. Which aligns with the previous ~310 seconds as the default value is `300.0` so a job retry seems to be delayed by the value of `job_timeout` + 10 seconds.

Is there any reason for this, or is it indeed a bug?

**Comment by Kamforka** [comment]
I investigated  further and the exception is raised in the worker's `close` method when running `await asyncio.gather(*self.tasks)`.

I also checked the worker's `run_job` method more precisely the block where the task function is being called and checked for an exception:

```python
            try:
                async with async_timeout.timeout(timeout_s):
                    result = await function.coroutine(ctx, *args, **kwargs)
            except Exception as e:
                exc_extra = getattr(e, "extra", None)
                if callable(exc_extra):
                    exc_extra = exc_extra()
                raise
```

The `function.coroutine` is called successfully, but in case of a Ctrl+C shutdown no exception is raised in the above block.

I also checked that the worker `handle_sig` method  is cancelling all running tasks, but the `CancelledError` exception never reaches the `run_job` method.

**Comment by sgranade** [comment]
It turns out that `except Exception as e` doesn't catch the `CancelledError`. I'm not sure why, precisely, since `CancelledError` inherits from `Exception`.

Changing it to `except (Exception, asyncio.CancelledError) as e` fixes the issue. I'll work up a PR with the fix.

**Comment by sgranade** [comment]
For future historians, the root of the problem is because, as of Python 3.8, [`CancelledError` inherits from `BaseException` rather than `Exception`](https://bugs.python.org/issue32528).

**Comment by samuelcolvin** [comment]
fixed by #213
